clang/test/CIR/CodeGen/switch-gnurange.cpp

```diff
@@ -0,0 +1,205 @@
+// LLVM-NEXT:      i32 0, label %[[CASE:[0-9]+]]
+// LLVM-NEXT:      i32 1, label %[[CASE]]
+// LLVM-NEXT:      i32 2, label %[[CASE]]
+// LLVM-NEXT:    ]
+//      LLVM:  [[CASE]]:
+//      LLVM:    store i32 1
+//      LLVM:    ret
+//      LLVM:  [[DEFAULT]]:
+//      LLVM:    store i32 0
+//      LLVM:    ret
+
+void sw3(enum letter c) {
+  int x = 0;
+  switch (c) { 
+  case A ... C:
+    x = 1;
+    break;
+  case D ... F:
+    x = 2;
+    break;
+  case G ... I:
+    x = 3;
+    break;
+  case J ... L:
+    x = 4;
+    break;
+  }
+}
+
+//      CIR:  cir.func @_Z3sw36letter
+//      CIR:    cir.scope {
+//      CIR:      cir.switch
+// CIR-NEXT:      case (anyof, [0, 1, 2] : !s32i) {
+// CIR-NEXT:        cir.int<1>
+//      CIR:        cir.break
+// CIR-NEXT:      },
+// CIR-NEXT:      case (anyof, [3, 4, 5] : !s32i) {
+// CIR-NEXT:        cir.int<2>
+//      CIR:        cir.break
+// CIR-NEXT:      },
+// CIR-NEXT:      case (anyof, [6, 7, 8] : !s32i) {
+// CIR-NEXT:        cir.int<3>
+//      CIR:        cir.break
+// CIR-NEXT:      },
+// CIR-NEXT:      case (anyof, [9, 10] : !s32i) {
+// CIR-NEXT:        cir.int<4>
+//      CIR:        cir.break
+// CIR-NEXT:      }
+// CIR-NEXT:      ]
+// CIR-NEXT:    }
+
+//      LLVM:  @_Z3sw36letter
+//      LLVM:    switch i32 %[[C:[0-9]+]], label %[[DEFAULT:[0-9]+]] [
+// LLVM-NEXT:      i32 0, label %[[CASE_AC:[0-9]+]]
+// LLVM-NEXT:      i32 1, label %[[CASE_AC]]
+// LLVM-NEXT:      i32 2, label %[[CASE_AC]]
+// LLVM-NEXT:      i32 3, label %[[CASE_DF:[0-9]+]]
+// LLVM-NEXT:      i32 4, label %[[CASE_DF]]
+// LLVM-NEXT:      i32 5, label %[[CASE_DF]]
+// LLVM-NEXT:      i32 6, label %[[CASE_GI:[0-9]+]]
+// LLVM-NEXT:      i32 7, label %[[CASE_GI]]
+// LLVM-NEXT:      i32 8, label %[[CASE_GI]]
+// LLVM-NEXT:      i32 9, label %[[CASE_JL:[0-9]+]]
+// LLVM-NEXT:      i32 10, label %[[CASE_JL]]
+// LLVM-NEXT:    ]
+//      LLVM:  [[CASE_AC]]:
+//      LLVM:    store i32 1, ptr %[[X:[0-9]+]]
+//      LLVM:    br label %[[EPILOG:[0-9]+]]
+//      LLVM:  [[CASE_DF]]:
+//      LLVM:    store i32 2, ptr %[[X]]
+//      LLVM:    br label %[[EPILOG]]
+//      LLVM:  [[CASE_GI]]:
+//      LLVM:    store i32 3, ptr %[[X]]
+//      LLVM:    br label %[[EPILOG]]
+//      LLVM:  [[CASE_JL]]:
+//      LLVM:    store i32 4, ptr %[[X]]
+//      LLVM:    br label %[[EPILOG]]
+//      LLVM:  [[EPILOG]]:
+//      LLVM:    ret void
+
+void sw4(int x) {
+  switch (x) {
+    case 66 ... 233:
+      break;
+    case -50 ... 50:
+      break;
+  }
+}
+
+//      CIR:  cir.func @_Z3sw4i
+//      CIR:    cir.scope {
+//      CIR:      cir.switch
+// CIR-NEXT:      case (anyof, [66, 67, 68, 69, {{[0-9, ]+}}, 230, 231, 232, 233] : !s32i) {
+// CIR-NEXT:        cir.break
+// CIR-NEXT:      },
+// CIR-NEXT:      case (anyof, [-50, -49, -48, -47, {{[0-9, -]+}}, -1, 0, 1, {{[0-9, ]+}}, 47, 48, 49, 50] : !s32i) {
+// CIR-NEXT:        cir.break
+// CIR-NEXT:      }
+// CIR-NEXT:      ]
+// CIR-NEXT:    }
+
+//      LLVM:  @_Z3sw4i
```

Comment:
For this switch, if we look at OG codegen, this is what we get instead: https://godbolt.org/z/Ydfznrzx5, it basically does a smart range check leading to smaller generated code. I'm fine with the current approach if you think you can work on another next PR that implements the desired behavior - one way to go about it is to add a `range` case kind and during lowering, you can the decide between `cir.switch.flat` or the smart range check.

Reply:
It looks reasonable, surprised I didn't notice it.
Of course, I'm happy to draft another pr later, will have a look on flat and lowering code firstly.

Reply:
Nice, thank you!